mavlink/qgroundcontrol
# GPS: Add support for GNSS receiver resilience information

Added support for GNSS receivers' resilience information

Description
-----------
This pull request modifies the UI to adapt to the resilience information for GPS interference (jamming and spoofing) as well as status for the GPS that were both recently added to mavlink. These changes only impact users that have GPS that use those mavlink messages, otherwise nothing new will show and it won't crowd the UI.

### Interference
For interference, the status are shown with a new icon that colors depending on the status:
| Status | Color | Example |
| ---  | --- | --- |
| Unknown (default) | None, the icon won't show | N/A |
| OK | White | ![image](https://github.com/user-attachments/assets/8482ef7c-eac0-453c-adfa-e3afc09bcf19) |
| Mitigated | Orange | ![image](https://github.com/user-attachments/assets/077d05c3-8d60-40ed-96af-61ee328d5a00) |
| Ongoing | Red | ![image](https://github.com/user-attachments/assets/8aa90a38-52b9-48ec-b653-65d6e1efc591) |

A drop-down menu will then allow user to have more insight into the kind of interference happening, for example:
| Example of mitigated interference | Example of interference |
| --- | --- |
| ![image](https://github.com/user-attachments/assets/d09fb930-19f8-49d0-a7bd-a6c5950f8b95) | ![image](https://github.com/user-attachments/assets/e8d4eea1-03f2-4c80-b963-b20dbae08224) |
Note: the lower-case 'i' has been fixed later

### Authentication
GNSS authentication is another way to check for interference, for example using Galileo OSNMA. An icon was also added for this purpose the same as the interference with a drop-down menu :
| Status | Color | Example |
| ---  | --- | --- |
| Unknown (default) | None, the icon won't show | N/A |
| OK | White |  |
| Initializing/In progress | Yellow | ![image](https://github.com/user-attachments/assets/c62689c4-cb83-4fc1-b557-cd1d20f9ab43) |
| Ongoing | Red | ![image](https://github.com/user-attachments/assets/7ab98c69-0fac-44ab-a96a-1f36cbc4ec40) |
| Available but sisabled | Gray | ![image](https://github.com/user-attachments/assets/de8e3a76-edb6-4457-a057-a500decca408) |

### GPS Status
The added GPS status message allow to know the error concerning the GPS. If one or more error happen, to warn the user, the GPS icon will turn red and a field, GPS status, will be added to the drop-down menu (it isn't visible otherwise).
| Status | Example |
| ---  | --- | 
| One error | ![image](https://github.com/user-attachments/assets/abb842ef-c70d-4b93-b54a-5b06dea81548) |
| Multiple errors | ![image](https://github.com/user-attachments/assets/2a3ab4f4-2596-4200-913e-2e8f64ca2371) | 

## NB
* I am more than willing to discuss any changes to the UI to improve this pull request, and implement them.

Test Steps
-----------
I used PX4 with the resilience message enabled with a Pixhawk4 and a Septentrio Mosaic go. Not all status could be easily recreated, interference was hard-coded to test its impact on the UI and not all error were tested, only ones that could be provoked.
As of creating this pull request, the [resilience PR](https://github.com/PX4/PX4-Autopilot/pull/23096) has to be approved after a minor style change and its logic will not be modified, it can be safely used for testing.

Checklist:
----------


Related Issue
-----------



By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
